src/combinatorics/catalan.lean

```diff
@@ -0,0 +1,126 @@
+  binomial coefficient, `catalan n = nat.central_binom n / (n + 1)`.
+
+## Implementation details
+
+The proof of `catalan_eq_central_binom_div` follows
+https://math.stackexchange.com/questions/3304415/catalan-numbers-algebraic-proof-of-the-recurrence-relation
+
+## TODO
+
+* Prove that the Catalan numbers enumerate many interesting objects.
+* Provide the many variants of Catalan numbers, e.g. associated to complex reflection groups,
+  Fuss-Catalan, etc.
+
+-/
+
+open_locale big_operators
+open finset
+
+/-- The recursive definition of the sequence of Catalan numbers:
+`catalan (n + 1) = ∑ i : fin n.succ, catalan i * catalan (n - i)` -/
+def catalan : ℕ → ℕ
+| 0       := 1
+| (n + 1) := ∑ i : fin n.succ, have _ := i.2, have _ := nat.lt_succ_iff.mpr (n.sub_le i),
+             catalan i * catalan (n - i)
+
+@[simp] lemma catalan_zero : catalan 0 = 1 := by rw catalan
+
+lemma catalan_succ (n : ℕ) : catalan (n + 1) = ∑ i : fin n.succ, catalan i * catalan (n - i) :=
+by rw catalan
+
+@[simp] lemma catalan_one : catalan 1 = 1 := by simp [catalan_succ]
+
+/-- A helper sequence that can be used to prove the equality of the recursive and the explicit
+definition using a telescoping sum argument. -/
+private def gosper_catalan (n j : ℕ) : ℚ :=
+nat.central_binom j * nat.central_binom (n - j) * (2 * j - n) / (2 * n * (n + 1))
+
+private lemma gosper_trick {n i : ℕ} (h : i ≤ n) :
+  gosper_catalan (n+1) (i+1) - gosper_catalan (n+1) i =
+  nat.central_binom i / (i + 1) * nat.central_binom (n - i) / (n - i + 1) :=
+begin
+  have : (n:ℚ) + 1 ≠ 0 := by exact_mod_cast n.succ_ne_zero,
+  have : (n:ℚ) + 1 + 1 ≠ 0 := by exact_mod_cast (n + 1).succ_ne_zero,
+  have : (i:ℚ) + 1 ≠ 0 := by exact_mod_cast i.succ_ne_zero,
+  have : (n:ℚ) - i + 1 ≠ 0 := by exact_mod_cast (n - i).succ_ne_zero,
+  have h₁ : ((i:ℚ) + 1) * (i + 1).central_binom = 2 * (2 * i + 1) * i.central_binom,
+  { exact_mod_cast nat.succ_mul_central_binom_succ i },
+  have h₂ : ((n:ℚ) - i + 1) * (n - i + 1).central_binom
+    = 2 * (2 * (n - i) + 1) * (n - i).central_binom,
+  { exact_mod_cast nat.succ_mul_central_binom_succ (n - i) },
+  simp only [gosper_catalan],
+  push_cast,
+  field_simp,
+  rw (nat.succ_sub h),
+  linear_combination
+    (2:ℚ) * (n - i).central_binom * (i + 1 - (n - i)) * (n + 1) * (n + 2) * ((n - i) + 1) * h₁
+    - 2 * i.central_binom * (n + 1) * (n + 2) * (i - (n - i) - 1) * (i + 1) * h₂,
+end
+
+private lemma gosper_catalan_sub_eq_central_binom_div (n : ℕ) :
+  gosper_catalan (n + 1) (n + 1) - gosper_catalan (n + 1) 0 = nat.central_binom (n + 1) / (n + 2) :=
+begin
+  have : (n:ℚ) + 1 ≠ 0 := by exact_mod_cast n.succ_ne_zero,
+  have : (n:ℚ) + 1 + 1 ≠ 0 := by exact_mod_cast (n + 1).succ_ne_zero,
+  have h : (n:ℚ) + 2 ≠ 0 := by exact_mod_cast (n + 1).succ_ne_zero,
+  simp only [gosper_catalan, nat.sub_zero, nat.central_binom_zero, nat.sub_self],
+  field_simp,
+  ring,
+end
+
+theorem catalan_eq_central_binom_div (n : ℕ) :
+  catalan n = nat.central_binom n / (n + 1) :=
```

Comment:
Could you please also include a lemma for the variant
```lean
  (n + 1) * catalan n = nat.central_binom n
```
?  I imagine this will also appear fairly often in applications.  Not sure whether it's quicker to prove this as a corollary of `catalan_eq_central_binom_div` or `catalan_eq_central_binom_div` as a corollary of it.